### DISCUSSION TITLE
Awaiting Approval?

### DISCUSSION THREAD

**Discussion by danielcmorris** [Automattic/wordpress-activitypub#200] [Q&A]
I installed this and have been trying it out.  I was able to find myself on mastodon and my entire profile was there:  EXTREMELY COOL.

@morrisdev@morrisdev.com

However, when I tried to follow the account, it says, "Awaiting Approval".  There's no place I can find to approve it.  Nobody in the followers list, no nothing.  Any hints would be appreciated.

**Comment by timnolte** [comment]
Are you running the LiteSpeed web server and/or ModSecurity? I had this problem and it was ModSecurity rules that were blocking some of the request types from being processed.

**Reply by danielcmorris** [reply]
Nope.  I disabled every plugin except ActivityPub and still no joy.  I'm trying to figure out where in the world the followers would go if they were to be approved, but there are no new tables or anything.  If it's using "users" maybe there's some kind of setting I'm missing.

**Reply by timnolte** [reply]
For me followers are listed under Profile->Followers (Fediverse). Also, it's important that you are logged into the right account. Each author/account on the WordPress site has it's own list of followers. When I'm logged into my personal account I see my followers but when I'm logged in with my admin account I don't see my personal account followers.

I will say though that I have never seen that you have to approve a follow in WordPress. I've only seen issues with following from a Mastodon account because of the WordPress web server blocking the types of requests being sent and not returning a correct response to the Mastodon instance. So I'm pretty sure you have a web server configuration issue not a WordPress/plugin issue.

**Reply by danielcmorris** [reply]
Unfortunately, it's just a generic HostGator site.  No server access.  Do you know the endpoint a follow request goes to?  I could test it out in Postman or something.

**Reply by timnolte** [reply]
If you load up the DevTools in Chrome, or the equivalent in your browser of choice, you should see the call that is being made when you click the Follow button.

**Reply by danielcmorris** [reply]
Unfortunately, the follow button is on my mastodon account, which posts back to the mastodon server and, I assume, the mastodon server then redirects it to my WordPress server.  So the post I see in Chrome tools doesn't help.  (I was sure that was the way to go!)

I guess I'll need to dig into the mastodon server code and see if I can find that endpoint.

**Reply by timnolte** [reply]
Ah, duh, you're right. So yeah, especially if it's not your own Mastodon instance you'll have to do some code digging. I have my own Mastodon instance and my own web server so I have access to all of the logs.

**Comment by enzocomics** [comment]
I had this issue after first installing as well. After digging through the support threads here and on WordPress.org, the solution that worked for me was adding  `application/activity+json` to the mime.types file in Apache (or Nginx). 

Assuming you're on shared hosting on Hostgator, you might have to contact them for help implementing that. If they offer any kind of cpanel, maybe you can check if you can edit MIME types there.

Edit: [Hostgator help article on adding a MIME type](https://www.hostgator.com/help/article/proper-mime-types)

**Reply by danielcmorris** [reply]
Well, that looks promising, but I added it and still got nothing.  However, my google search using that activity+json turned up this:

https://blog.rac.me.uk/tag/owasp3/

Which says I need to add a new mod security rule to allow application/activity+json.  And, it seems I need root access to do that.  I actually do have WHM access, but it doesn't include all the additional features this guy has.  So, I guess I'm going to have to contact my admin, but maybe I'll just put it on a machine I actually  have root access to.  

Thanks for the help.

**Reply by enzocomics** [reply]
Please let us know how it goes!

**Comment by craigconstantine** [comment]
I have an account on a mastodon server. Did a follow, and I see my account listed under `Followers > Fediverse` in my Wordpress admin area…

<img width="490" alt="Screenshot 2022-12-02 at 12 10 04" src="https://user-images.githubusercontent.com/2278992/205347214-de6b3143-2b6a-4ade-8f27-8f49746bc54c.png">

But over on the mastodon server, I still see it as a pending request…

<img width="590" alt="Screenshot 2022-12-02 at 12 11 32" src="https://user-images.githubusercontent.com/2278992/205347588-19d47d84-75c7-4362-9347-980468757e06.png">

¿ Does this mean my Wordpress server's response wasn't understood by the `c.im` mastodon server where my account is ?

¿ Is there a way to `curl` (unix sysadmin here, can just point me to howto) a follow request ?

**Reply by craigconstantine** [reply]
For example, this works fine. But I don't know how to do a follow-request action…

```
$ curl -i 'https://constantine.name/.well-known/webfinger?resource=acct:craig@constantine.name'
HTTP/2 200
x-robots-tag: noindex
link: <https://constantine.name/wp-json/>; rel="https://api.w.org/"
x-content-type-options: nosniff
access-control-expose-headers: X-WP-Total, X-WP-TotalPages, Link
access-control-allow-headers: Authorization, X-WP-Nonce, Content-Disposition, Content-MD5, Content-Type, X-HTTP-Method-Override
allow: GET
vary: Origin,Accept-Encoding
content-type: application/json; charset=UTF-8
date: Sat, 03 Dec 2022 12:58:07 GMT
server: Apache

{"subject":"craig@constantine.name","aliases":["https:\/\/constantine.name\/author\/craig\/"],"links":[{"rel":"self","type":"application\/activity+json","href":"https:\/\/constantine.name\/author\/craig\/"},{"rel":"http:\/\/webfinger.net\/rel\/profile-page","type":"text\/html","href":"https:\/\/constantine.name\/author\/craig\/"}]}
```

**Reply by craigconstantine** [reply]
I did some debugging, and my problem is the same as https://github.com/pfefferle/wordpress-activitypub/issues/192  …I'll stop writing in this discussion :)
